cmd/list.go

```diff
@@ -0,0 +1,180 @@
+package cmd
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/Benchkram/errz"
+	"github.com/spf13/cobra"
+
+	"github.com/puppetlabs/wash/api/client"
+	apitypes "github.com/puppetlabs/wash/api/types"
+	cmdutil "github.com/puppetlabs/wash/cmd/util"
+	"github.com/puppetlabs/wash/config"
+	"github.com/puppetlabs/wash/plugin"
+)
+
+func listCommand() *cobra.Command {
+	listCmd := &cobra.Command{
+		Use:     "list [file]",
+		Aliases: []string{"ls"},
+		Short:   "Lists the resources at the indicated path.",
+		Args:    cobra.MaximumNArgs(1),
+	}
+
+	listCmd.RunE = toRunE(listMain)
+
+	return listCmd
+}
+
+func headers(timey string) []cmdutil.ColumnHeader {
+	return []cmdutil.ColumnHeader{
+		{ShortName: "name", FullName: "NAME"},
+		{ShortName: "time", FullName: timey},
+		{ShortName: "verbs", FullName: "ACTIONS"},
+	}
+}
+
+func format(t time.Time) string {
+	return t.Format(time.RFC822)
+}
+
+func formatListEntries(apiPath string, ls []apitypes.ListEntry) string {
+	table := make([][]string, len(ls))
+	for i, entry := range ls {
+		var ctimeStr string
+		if entry.Attributes.HasCtime() {
+			ctimeStr = format(entry.Attributes.Ctime())
+		} else {
+			ctimeStr = "<unknown>"
+		}
+
+		verbs := strings.Join(entry.Actions, ", ")
+
+		name := entry.CName
+		if len(ls) > 1 && entry.Path == apiPath {
+			// Represent the pwd as "."
+			name = "."
+		}
+
+		if isListable(entry) {
+			name += "/"
+		}
+
+		table[i] = []string{name, ctimeStr, verbs}
+	}
+	return cmdutil.FormatTable(headers("CREATED"), table)
+}
+
+func findEntry(entries []apitypes.ListEntry, name string) apitypes.ListEntry {
+	for _, entry := range entries {
+		if entry.CName == name {
+			return entry
+		}
+	}
+	return apitypes.ListEntry{}
+}
+
+func isListable(entry apitypes.ListEntry) bool {
+	for _, action := range entry.Actions {
+		if action == plugin.ListAction.Name {
+			return true
+		}
+	}
+	return false
+}
+
+func listResource(apiPath string) error {
+	conn := client.ForUNIXSocket(config.Socket)
+
+	var entries []apitypes.ListEntry
+	if apiPath == "/" {
+		// The root, definitely listable
+		ls, err := conn.List(apiPath)
+		if err != nil {
+			return err
+		}
+		entries = ls
+	} else {
+		// List the parent to see whether it's a single entry or a listable resource
+		parent, base := filepath.Split(apiPath)
+		parentEntries, err := conn.List(parent)
+		if err != nil {
+			return err
+		}
+
+		target := findEntry(parentEntries, base)
+		if target.CName != base || isListable(target) {
+			// If we didn't find a parent entry, just try listing it. Can happen if the type has changed
+			// or disappeared, and List will give a reasonable error in that case.
+			ls, err := conn.List(apiPath)
+			if err != nil {
+				return err
+			}
+			entries = ls
+		} else {
+			entries = []apitypes.ListEntry{target}
+		}
+	}
+
+	fmt.Print(formatListEntries(apiPath, entries))
+	return nil
+}
+
+func listPath(path string) error {
```

Comment:
One thing I was thinking about is to make a core plugin that we can delegate to for "non-Wash" paths (this core-plugin's internal and is not something we expose to the user, but we include it in the plugin.Registry). This way, we can maintain consistency between the outputs in `wash ls` for Wash entries vs. non-Wash entries.

FS-specific things like inode number can be included in the metadata so that `wash find` could still query regular files via the `meta` primary.

What do you think? Asking b/c having commands with two different outputs based on context is a little weird. There's no reason why regular files cannot be interpreted as Wash resources (especially when we have an `sshfs` example).

Reply:
Hmm, we could maybe make it an exported variable instead of adding it to the plugin.Registry.

Reply:
This "plugin" could probably just be in the API, b/c it's needed for the API to work on regular files. That also means our endpoints need to take absolute paths.

Reply:
I think the endpoints may need to take absolute paths anyway to support Windows. I'd prefer to keep that a separate refactor, as I think it threads through a lot more of the system.

Reply:
Sounds good.